contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -106,13 +101,88 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+Container_ using the API-level concept of an _Ephemeral Container_. The API
+doesn't prescribe how an Ephemeral Container is used. It could conceivably see
+use other than Debug Containers, but we don't currently have other use cases.
+
+#### Chosen Solution: POST to an Existing Pod
+
+We're modifying an existing pod, so this fits as a subresource of the target
+pod. We will create a new top-level object that contains a `v1.Container` and
+`POST` to the subresource to create the Debug Container. Since this is a `POST`,
+we cannot upgrade the connection to streaming if we want to continue supporting
+web socket clients.
+
+Rather than using an existing subresource like `/exec` and conditionally stream
+based on `PodExecOptions`, we will create a new subresource with consistent
+streaming behavior. A new subresource has the added benefit of being able to
+entirely hide interface entirely behind a feature flag by conditionally
+registering the new subresource.
+
+A `v1.Container` by itself lacks type and object metadata, so we will create a
+new type:
+
+```
+// EphemeralContainer describes a container to attach to a running pod for troubleshooting.
+type EphemeralContainer struct {
+        metav1.TypeMeta `json:",inline"`
+
+        // Spec describes the Ephemeral Container to be created.
+        Spec *Container `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+        // Most recently observed status of the container.
+        // This data may not be up to date.
+        // Populated by the system.
+        // Read-only.
+        // +optional
+        Status *ContainerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+
+        // If set, the name of the container from PodSpec that this ephemeral container targets.
+        // If not set then the ephemeral container is run in whatever namespaces are shared
+        // for the pod.
+        TargetContainerName string `json:"targetContainerName,omitempty" protobuf:"bytes,4,opt,name=targetContainerName"`
+}
+```
 
-We will extend `v1.Pod`'s `/exec` subresource to support "executing" container
-images. The current `/exec` endpoint must implement `GET` to support streaming
-for all clients. We don't want to encode a (potentially large) `v1.Container` as
-an HTTP parameter, so we must extend `v1.PodExecOptions` with the specific
-fields required for creating a Debug Container:
+**Note that Ephemeral Containers are not regular containers and should not be
+used to build services.** They lack guarantees for resources or execution, and
+many of the fields of `v1.Container` will not be allowed for Debug Containers. A
+request will be rejected if any field is set other than the following
+whitelisted fields: `Name`, `Image`, `Command`, `Args`, `WorkingDir`, `Env`,
+`EnvFrom`, `ImagePullPolicy`, `SecurityContext`. `TTY` and `Stdin` are always
+enabled for Debug Containers and will be ignored.
+
+The new `/add` subresource allows the following:
```

Comment:
is the subresource `/add` or `/ephemeralContainers`? being more explicit seems better (though we don't mix case in resource or subresource names)

Reply:
It's `/ephemeralcontainers` (I discovered the problem with case on subresources today, too). Fixed.

The kubelet API seems to still mix case, so I left that one alone. sig-node probably has a preference but I'm not sure what it is.